raiden/storage/sqlite.py

```diff
@@ -0,0 +1,247 @@
+
+    if result[0] != data:
+        raise RuntimeError(
+            'Database cannot save NUL character, ensure python is at least 2.7.3'
+        )
+
+    conn.rollback()
+
+
+class SQLiteStorage:
+    def __init__(self, database_path, serializer):
+        conn = sqlite3.connect(database_path)
+        conn.text_factory = str
+        conn.execute('PRAGMA foreign_keys=ON')
+
+        with conn:
+            cursor = conn.cursor()
+            cursor.execute(
+                'CREATE TABLE IF NOT EXISTS state_changes ('
+                '    identifier INTEGER PRIMARY KEY AUTOINCREMENT, '
+                '    data BINARY'
+                ')'
+            )
+            cursor.execute(
+                'CREATE TABLE IF NOT EXISTS state_snapshot ('
+                '    identifier INTEGER PRIMARY KEY, '
+                '    statechange_id INTEGER, '
+                '    data BINARY, '
+                '    FOREIGN KEY(statechange_id) REFERENCES state_changes(identifier)'
+                ')'
+            )
+            cursor.execute(
+                'CREATE TABLE IF NOT EXISTS state_events ('
+                '    identifier INTEGER PRIMARY KEY, '
+                '    source_statechange_id INTEGER NOT NULL, '
+                '    block_number INTEGER NOT NULL, '
+                '    data BINARY, '
+                '    FOREIGN KEY(source_statechange_id) REFERENCES state_changes(identifier)'
+                ')'
+            )
+
+        sqlite_sanity_check(conn)
+
+        # When writting to a table where the primary key is the identifier and we want
+        # to return said identifier we use cursor.lastrowid, which uses sqlite's last_insert_rowid
+        # https://github.com/python/cpython/blob/2.7/Modules/_sqlite/cursor.c#L727-L732
+        #
+        # According to the documentation (http://www.sqlite.org/c3ref/last_insert_rowid.html)
+        # if a different thread tries to use the same connection to write into the table
+        # while we query the last_insert_rowid, the result is unpredictable. For that reason
+        # we have this write lock here.
+        #
+        # TODO (If possible):
+        # Improve on this and find a better way to protect against this potential race
+        # condition.
+        self.write_lock = threading.Lock()
+        self.conn = conn
+        self.serializer = serializer
+
+    def write_state_change(self, state_change):
+        serialized_data = self.serializer.serialize(state_change)
+
+        with self.write_lock, self.conn:
+            cursor = self.conn.execute(
+                'INSERT INTO state_changes(identifier, data) VALUES(null, ?)',
+                (serialized_data,)
+            )
+            last_id = cursor.lastrowid
+
+        return last_id
+
+    def write_state_snapshot(self, statechange_id, snapshot):
+        # TODO: Snapshotting is not yet implemented. This is just skeleton code
+        # Issue: https://github.com/raiden-network/raiden/issues/593
```

Comment:
this issue is closed: https://github.com/raiden-network/raiden/issues/593
what's the status of this todo?

Reply:
https://github.com/raiden-network/raiden/issues/682